### PR TITLE
header requesthandler added

### PR DESCRIPTION
**Pull request by younger15**

### CODE REVIEW

**Comment by younger15** [comment]
code style changed

File: src/Header/RequestHandler.h [hunk 1]
```diff
@@ -0,0 +1,25 @@
+// Copyright (c) 2021, 鍾淯丞, 周杰仕, 林仁鴻. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#include "./OperateData.h"
```
Comment (john0312) [inline src/Header/RequestHandler.h]: Please add include path in cmake, so you don't have to add "./" to your include directive.

File: src/Header/RequestHandler.h [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+//Will decode the request string and check if it's a valid request.
+//calling RequestParser to get OperateData then call correspond api function
+class RequestHandler
+{
+
+public:
+    //default constructor will never be used
+    RequestHandler();
+    //instantiate instance with request string, will parse to OperateData and check with Authenticator and call correspond api.
+    RequestHandler(const string &requestData);
```
Comment (john0312) [inline src/Header/RequestHandler.h]: You might need to #include <string> for this to work.
Also, you should reference either reference it by its full path (std::string), or put "using std::string" in your class declaration.

File: src/Header/RequestHandler.h [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+    //data is content of request body
+    unique_ptr<OperateData> data;
```
Comment (john0312) [inline src/Header/RequestHandler.h]: You need to include <memory> for this. Also, reference it as std::unique_ptr.

File: src/Header/RequestHandler.h [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+    //header is content of request head
+    unique_ptr<OperateData> header;
+    //storing all feature api
+    std::unordered_map<std::string, APIHandler> apis;
```
Comment (john0312) [inline src/Header/RequestHandler.h]: Please add include.

File: src/Header/RequestHandler.h [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+};
```
Comment (john0312) [inline src/Header/RequestHandler.h]: Please add new line.

**Comment by john0312** [comment]
Hello @mike820324 (Ruei-Min Jiang), please take a look at this commit/PR. Thank you.

File: src/Header/RequestHandler.h (hunk 2 above)
Comment (younger15) [reply src/Header/RequestHandler.h]: Done

File: src/Header/RequestHandler.h (hunk 4 above)
Comment (younger15) [reply src/Header/RequestHandler.h]: Done

File: src/Header/RequestHandler.h (hunk 3 above)
Comment (younger15) [reply src/Header/RequestHandler.h]: Done

File: src/Header/RequestHandler.h (hunk 5 above)
Comment (younger15) [reply src/Header/RequestHandler.h]: Done

File: src/Header/RequestHandler.h [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+#include <unordered_map>
+#include <memory>
+
+//Will decode the request string and check if it's a valid request.
+//calling RequestParser to get OperateData then call correspond api function
+class RequestHandler
+{
+public:
+    //default constructor will never be used
+    RequestHandler();
```
Comment (john0312) [inline src/Header/RequestHandler.h]: If it's not used, then please delete it.

File: src/Header/RequestHandler.h [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+    //instantiate instance with request string, will parse to OperateData and check with Authenticator and call correspond api.
+    RequestHandler(const std::string &requestData);
+    //data is content of request body
+    std::unique_ptr<OperateData> data;
```
Comment (john0312) [inline src/Header/RequestHandler.h:26]: If I understand correctly, OperateData represents the entire HTTP request. If that's the case, you probably don't need to separate data from header.

File: src/Header/RequestHandler.h [hunk 8]
```diff
@@ -0,0 +1,32 @@
...
+    //header is content of request head
+    std::unique_ptr<OperateData> header;
+    //storing all feature api
+    std::unordered_map<std::string, APIHandler> apis;
```
Comment (john0312) [inline src/Header/RequestHandler.h]: Consider making "apis" static const, it shouldn't be changed and there should only be one copy.

File: src/Header/RequestHandler.h [hunk 9]
```diff
@@ -0,0 +1,32 @@
...
+
+//Will decode the request string and check if it's a valid request.
+//calling RequestParser to get OperateData then call correspond api function
+class RequestHandler
+{
+public:
+    //default constructor will never be used
+    RequestHandler();
+    //instantiate instance with request string, will parse to OperateData and check with Authenticator and call correspond api.
+    RequestHandler(const std::string &requestData);
```
Comment (john0312) [inline src/Header/RequestHandler.h:24]: Do you want to take a string or a socket for HTTP? If you want to take a string, you'll need something else to read the request out of the socket.

File: src/Header/RequestHandler.h (hunk 1 above)
Comment (younger15) [reply src/Header/RequestHandler.h]: Done

File: src/Header/RequestHandler.h (hunk 6 above)
Comment (younger15) [reply src/Header/RequestHandler.h]: Done

File: src/Header/RequestHandler.h (hunk 8 above)
Comment (younger15) [reply src/Header/RequestHandler.h]: Done

File: src/Header/RequestHandler.h (hunk 7 above)
Comment (younger15) [reply src/Header/RequestHandler.h:26]: Done

File: src/Header/RequestHandler.h [hunk 10]
```diff
@@ -0,0 +1,28 @@
...
+//calling RequestParser to get OperateData then call correspond api function
+class RequestHandler
+{
+public:
+    //instantiate instance with request, will parse to OperateData and check with Authenticator and call correspond api.
+    RequestHandler(const byte &requestData);
+    //content of request
+    std::unique_ptr<OperateData> data;
+    //storing all feature api
+    std::unordered_map<std::string, static const APIHandler&> apis;
```
Comment (john0312) [inline src/Header/RequestHandler.h]: static const std::unordered_map<std::string, APIHandler> apis;
You can initialize it with an initializer list.

File: src/Header/RequestHandler.h [hunk 11]
```diff
@@ -0,0 +1,28 @@
...
+};
+
+#endif
```
Comment (john0312) [inline src/Header/RequestHandler.h]: Please add new line.

File: src/Header/RequestHandler.h [hunk 12]
```diff
@@ -0,0 +1,28 @@
+// Copyright (c) 2021, 鍾淯丞, 周杰仕, 林仁鴻. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#ifndef GROUP4_HEADER_REQUESTHANDLER_H_
+#define GROUP4_HEADER_REQUESTHANDLER_H_
+
+#include "OperateData.h"
```
Comment (john0312) [inline src/Header/RequestHandler.h:13]: Please sort and order your includes according to the guide:
https://google.github.io/styleguide/cppguide.html#Names_and_Order_of_Includes

File: src/Header/RequestHandler.h [hunk 13]
```diff
@@ -0,0 +1,28 @@
...
+#include <unordered_map>
+#include <memory>
+
+//Will decode the request string and check if it's a valid request.
+//calling RequestParser to get OperateData then call correspond api function
+class RequestHandler
+{
+public:
+    //instantiate instance with request, will parse to OperateData and check with Authenticator and call correspond api.
+    RequestHandler(const byte &requestData);
```
Comment (john0312) [inline src/Header/RequestHandler.h]: If you want to take a socket:
- If you're using asio, you'll take this: https://www.boost.org/doc/libs/1_66_0/doc/html/boost_asio/reference/ip__tcp/socket.html
- If you are using raw unix socket, you'll take a file descriptor (int).

File: src/Header/RequestHandler.h (hunk 9 above)
Comment (younger15) [reply src/Header/RequestHandler.h:24]: String will be read from socket in httpListener.

File: src/Header/RequestHandler.h (hunk 12 above)
Comment (younger15) [reply src/Header/RequestHandler.h:13]: Done

File: src/Header/RequestHandler.h [hunk 14]
```diff
@@ -0,0 +1,30 @@
+// Copyright (c) 2021, 鍾淯丞, 周杰仕, 林仁鴻. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#ifndef GROUP4_HEADER_REQUESTHANDLER_H_
+#define GROUP4_HEADER_REQUESTHANDLER_H_
+
+#include <string>
```
Comment (john0312) [inline src/Header/RequestHandler.h:10]: Please sort include order.
https://google.github.io/styleguide/cppguide.html#Names_and_Order_of_Includes

File: src/Header/RequestHandler.h (hunk 13 above)
Comment (younger15) [reply src/Header/RequestHandler.h]: request will be string from file descriptor

File: src/Header/RequestHandler.h (hunk 14 above)
Comment (younger15) [reply src/Header/RequestHandler.h:10]: Done

File: src/Header/RequestHandler.h (hunk 11 above)
Comment (younger15) [reply src/Header/RequestHandler.h]: Done

File: src/Header/RequestHandler.h (hunk 10 above)
Comment (younger15) [reply src/Header/RequestHandler.h]: Done

**Comment by john0312** [comment]
LGTM, approved.

**Pull request merged**